### PR TITLE
[FLINK-33354][runtime] Cache TaskInformation and JobInformation to avoid deserializing duplicate big objects

### PR DESCRIPTION
**Pull request by 1996fanrui**
## What is the purpose of the change

The background is similar to [FLINK-33315](https://issues.apache.org/jira/browse/FLINK-33315).

A hive table with a lot of data, and the HiveSource#partitionBytes is 281MB. When slotPerTM = 4, one TM will run 4 HiveSources at the same time.

How the TaskExecutor to submit a large task?

1. TaskExecutor#loadBigData will read all bytes from file to SerializedValue<TaskInformation> 
  - The SerializedValue<TaskInformation>  has a byte[]
  - It will cost the heap memory
  - It will be great than 281 MB, because it not only stores HiveSource#partitionBytes, it also stores other information of TaskInformation.
2. Generate the TaskInformation from SerializedValue<TaskInformation> 
  - TaskExecutor#submitTask calls the tdd.getSerializedTaskInformation()..deserializeValue()
  - tdd.getSerializedTaskInformation() is SerializedValue<TaskInformation> 
  - It will generate the TaskInformation
  - TaskInformation includes the Configuration taskConfiguration
  - The taskConfiguration includes StreamConfig#SERIALIZEDUDF

Based on the above process, TM memory will have 2 big byte array for each task:

- The SerializedValue<TaskInformation>
- The TaskInformation

When one TM runs 4 HiveSources at the same time, it will have 8 big byte array.

In our production environment, this is also a situation that often leads to TM OOM.

## Brief change log

- [FLINK-33354][runtime][refactor] Refactor ShuffleDescriptorsCache into a generic GroupCache
- [FLINK-33354][runtime][refactor] serializedJobInformation and taskInfo are never null
- [FLINK-33354][runtime] Cache TaskInformation and JobInformation to avoid deserializing duplicate big objects
- [FLINK-33354][runtime] Using the InputStream instead of byte array to avoid contiguous huge memory usage

## Verifying this change

Improve the old tests:

- DefaultGroupCacheTest
- TaskDeploymentDescriptorTest

Add a new test: `DefaultGroupCacheTest#testTaskInformationCache`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency):  no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature?no

### CODE REVIEW

**Comment by 1996fanrui** [comment]
Hi @pnowojski @RocMarshal , would you mind helping take a look this PR in your free time? Thanks a lot

**Comment by 1996fanrui** [comment]
Hi @huwh , would you mind helping take a look this PR in your free time as well?

This improvement is totally similar with FLINK-32386 is contributed by you, and this PR refactor the ShuffleDescriptorsCache into a generic GroupCache, so it would be better if you join this review, thanks a lot!

File: flink-runtime/src/main/java/org/apache/flink/runtime/util/DefaultGroupCache.java [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+            cachedBlobKeysPerJob.computeIfPresent(
+                    cacheKey.getGroup(),
+                    (group, keys) -> {
+                        keys.remove(cacheKey);
+                        if (keys.isEmpty()) {
+                            return null;
+                        } else {
+                            return keys;
+                        }
+                    });
```
Comment (RocMarshal) [inline flink-runtime/src/main/java/org/apache/flink/runtime/util/DefaultGroupCache.java:99]: Would be there a risk of memory leakage here?
For example, let's talk about the situation:

- There are too many groups
- Perform the following operations on each of these groups one by one
  - Add a set for one group and then remove set for the same one group, but the key has not been removed. Would there be many Entries in the form of `Entry-i<Group-i, set-i>`(set-i is empty or null) ?
  
In short, would `cachedBlobKeysPerJob` degenerate into a collection with too many elements?

Please correct me if needed for my limited read.

File: flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptor.java [hunk 2]
```diff
@@ -151,39 +166,43 @@ public TaskDeploymentDescriptor(
...
             NonOffloaded<TaskInformation> taskInformation =
                     (NonOffloaded<TaskInformation>) serializedTaskInformation;
-            return taskInformation.serializedValue;
-        } else {
-            throw new IllegalStateException(
-                    "Trying to work with offloaded serialized job information.");
+            return taskInformation.serializedValue.deserializeValue(getClass().getClassLoader());
         }
+        throw new IllegalStateException(
+                "Trying to work with offloaded serialized task information.");
```
Comment (RocMarshal) [inline flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptor.java:205]: How about 
```
        Preconditions.checkState(
                serializedJobInformation instanceof NonOffloaded,
                 "Trying to work with offloaded serialized job information.");
        NonOffloaded<JobInformation> jobInformation =
                (NonOffloaded<JobInformation>) serializedJobInformation;
        return jobInformation.serializedValue.deserializeValue(getClass().getClassLoader());


```

?

**Review by RocMarshal — COMMENTED** [review]
Thanks @1996fanrui for the contribution.
I left a few of comments.
Please let me know what's your opinion~  :)

File: flink-runtime/src/main/java/org/apache/flink/runtime/util/DefaultGroupCache.java (hunk 1 above)
Comment (1996fanrui) [reply flink-runtime/src/main/java/org/apache/flink/runtime/util/DefaultGroupCache.java:99]: I write a demo on My IDEA, it doesn't have the memory leak. When return null, the map will remove the key.


image


image

File: flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptor.java [hunk 3]
```diff
@@ -253,13 +274,19 @@ public void loadBigData(
...
+                // NOTE: Do not delete the job info BLOB since it may be needed again during
+                // recovery. (it is deleted automatically on the BLOB server and cache when the job
+                // enters a terminal state)
+                jobInformation =
+                        InstantiationUtil.deserializeObject(
+                                new BufferedInputStream(Files.newInputStream(dataFile.toPath())),
+                                getClass().getClassLoader());
+                jobInformationCache.put(jobId, jobInfoKey, jobInformation);
+            }
+            this.jobInformation = jobInformation.deepCopy();
```
Comment (huwh) [inline flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptor.java:289]: Can we use `this.jobInformation = jobInformation` here ?

**Review by huwh — APPROVED** [review]
Thanks @1996fanrui for preparing this PR.  LGTM with a minor comment.

File: flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptor.java (hunk 3 above)
Comment (huwh) [reply flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptor.java:289]: Thanks for your explanation.  Let's keep the `deepCopy`.

**Review by RocMarshal — APPROVED** [review]
Thanks for the update & review.
LGTM +1.

**Review by 1996fanrui — COMMENTED** [review]
Thanks for the review, merging~

**Pull request merged**